### PR TITLE
fsmap: switch to using iterator based loop

### PR DESCRIPTION
**Pull request by aliakseimakarau**
Segfault was triggered by the fs/mirror test.
The problem is in potential undefined behaviour
after erasing element and iterator invalidation in
FSMap::filter().
Fixed by explicit iterator increment.

Fixes: https://tracker.ceph.com/issues/55134

Signed-off-by: Aliaksei Makarau <aliaksei.makarau@ibm.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by ljflores** [comment]
@aliakseimakarau thanks for the proposed fix! Please change the PR so this merges to the master branch, and then we can backport it to pacific (I wrote a bit how to do this in the email I sent to you).

File: src/msg/msg_types.h [hunk 1]
```diff
@@ -543,12 +547,67 @@ WRITE_CLASS_ENCODER_FEATURES(entity_addr_t)
...
 std::ostream& operator<<(std::ostream& out, const entity_addr_t &addr);
 
-inline bool operator==(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) == 0; }
-inline bool operator!=(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) != 0; }
-inline bool operator<(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) < 0; }
-inline bool operator<=(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) <= 0; }
-inline bool operator>(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) > 0; }
-inline bool operator>=(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) >= 0; }
+inline bool operator==(const entity_addr_t& a, const entity_addr_t& b) {
+  struct sockaddr_storage ssa = a.get_sockaddr_storage();
```
Comment (ronen-fr) [inline src/msg/msg_types.h:551]: Is this include used in 'C' files? if not - no need for the 'struct'

**Comment by ronen-fr** [comment]
@aliakseimakarau : a question: the existing code erases the x.first, not 'x'. Why is this an undefined behaviour?
Tx

**Comment by aliakseimakarau** [comment]
> @aliakseimakarau : a question: the existing code erases the x.first, not 'x'. Why is this an undefined behaviour? Tx

Hi @ronen-fr,
we are iterating a container and modifying it, in general this should invalidate the iterators.
The erasure of the current element will invalidate the iterator at the position, therefore we have to jump to the next element: `it = filesystems.erase(it);`.

**Comment by ronen-fr** [comment]
> Hi @ronen-fr,
> we are iterating a container and modifying it, in general this should invalidate the iterators.
> The erasure of the current element will invalidate the iterator at the position, therefore we have to jump to the next element: `it = filesystems.erase(it);`.

Sorry - I didn't notice that we are actually removing an element from the same container. You are correct, of course.

**Comment by aliakseimakarau** [comment]
> > Hi @ronen-fr,
> > we are iterating a container and modifying it, in general this should invalidate the iterators.
> > The erasure of the current element will invalidate the iterator at the position, therefore we have to jump to the next element: `it = filesystems.erase(it);`.
> 
> Sorry - I didn't notice that we are actually removing an element from the same container. You are correct, of course.

Dear @ronen-fr,
thank you very much for checking!
I would like to say that a similar issue is in the `std::vector<mds_gid_t> FSMap::stop(mds_gid_t who)` but this issue i resolve in the next PR (the `fs:multifs` fails in this method).

File: src/msg/msg_types.h (hunk 1 above)
Comment (aliakseimakarau) [reply src/msg/msg_types.h:551]: > Is this include used in 'C' files? if not - no need for the 'struct'

Let me try to rebuild ceph with and without the struct keyword and find if it is included :-)

Comment (aliakseimakarau) [reply src/msg/msg_types.h:551]: Here was a discussion, therefore was found that all the struct info should be used for comparison/sorting
[https://github.com/ceph/ceph/pull/45095](https://github.com/ceph/ceph/pull/45095)

Comment (aliakseimakarau) [reply src/msg/msg_types.h:551]: > Is this include used in 'C' files? if not - no need for the 'struct'

The struct keyword can be removed.

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by aliakseimakarau** [comment]
Closing #45757 in favor of #45778.

**Pull request closed without merging**